### DISCUSSION TITLE
Snapcast Client for Arduino...

### DISCUSSION THREAD

**Discussion by pschatzmann** [snapcast/snapcast#1171] [Show and tell]
I thought it might be cool to have a Snapcast Client for __Arduino__!

I finally managed to have a first [DRAFT version](https://github.com/pschatzmann/arduino-snapclient) running that integrates nicely with my [Arduino Audio Tools](https://github.com/pschatzmann/arduino-audio-tools) project.

I still need to work a bit on the synchronization and it should be quite easy to extend it to support additional codecs...

**Comment by luar123** [comment]
Looks very interesting, thanks for sharing. How good does your synchronization work? I think this was always an issue with the original implementation (and RAM usage).
There is also this actively developed version https://github.com/CarlosDerSeher/snapclient where synchronization works really well.

**Comment by pschatzmann** [comment]
RAM is not a problem in my implementation, because I do not store any PCM data.

I have just added the synchronization starting logic today and it looks pretty promising: it can be fine tuned with the CONFIG_PLAYBACK_LAG_MS parameter because this might depend on the selected output (e.g I2S buffer size) and  codec.

As far as the fine tuning of the playback speed is concerned, I am not sure what the best approach would be:  I am currently thinking about using some dynamic resampling. It would need to be a mechanism which is generally available on the different Arduino platforms and has a minimal impact on the needed RAM.

I have seen that Carlos is using a timer and some filters for this, so this is not really an option...

Any input will be appreciated!
